ams/ams-server/src/main/java/com/netease/arctic/ams/server/service/impl/OptimizeQueueService.java

```diff
@@ -484,117 +484,137 @@ public String queueName() {
+        }
+      } else {
+        return onExecuteOptimizeTask(task, jobId, attemptId);
+      }
+    }
+
+    private OptimizeTask onExecuteOptimizeTask(OptimizeTaskItem task, JobId jobId, String attemptId) {
+      TableTaskHistory tableTaskHistory;
+      try {
+        // load files from sysdb
+        task.setFiles();
+        // update max execute time
+        task.setMaxExecuteTime();
+        tableTaskHistory = task.onExecuting(jobId, attemptId);
+      } catch (Exception e) {
+        task.clearFiles();
+        LOG.error("{} handle sysdb failed, try put task back into queue", task.getTaskId(), e);
+        if (!tasks.offer(task)) {
+          LOG.error("{} failed to put task back into queue", task.getTaskId());
+          task.onFailed(new ErrorMessage(System.currentTimeMillis(), "failed to put task back into queue"), 0);
+        }
+        throw e;
+      }
+      try {
+        insertTableTaskHistory(tableTaskHistory);
+      } catch (Exception e) {
+        LOG.error("failed to insert tableTaskHistory, {} ignore", tableTaskHistory, e);
+      }
+      return task.getOptimizeTask();
+    }
+
+    private OptimizeTaskItem waitForTask(long startTime, long waitTime) {
       while (true) {
+        lock();
+        try {
+          // if timeout, return null, await support zero or a negative value, and return false
+          if (planThreadCondition.await(waitTime - (System.currentTimeMillis() - startTime),
+              TimeUnit.MILLISECONDS)) {
+            OptimizeTaskItem task = tasks.poll();
+            if (task != null) {
+              if (!tables.contains(task.getTableIdentifier())) {
+                LOG.warn("get task {} from queue {} but table {} not in this queue",
+                    task.getTaskId(), queueName(), task.getTableIdentifier());
+                continue;
+              }
+              return task;
+            }
+          } else {
+            LOG.debug("The queue {} has no task have planned", optimizeQueue.getOptimizeQueueMeta().getQueueId());
+            return null;
+          }
+        } catch (InterruptedException e) {
+          LOG.warn("await for task was interrupted, return null", e);
+          return null;
+        } finally {
+          unlock();
+        }
         long duration = System.currentTimeMillis() - startTime;
         if (duration > waitTime) {
-          LOG.warn("pool task cost too much time {} ms, return null", duration);
+          LOG.info("pool task cost too much time {} ms, return null", duration);
           return null;
         }
-        OptimizeTaskItem task = tasks.poll();
-        if (task == null) {
+      }
+    }
+
+    private void tryPlanAsync(JobId jobId, String attemptId) {
+      if (!planThreadStarted.compareAndSet(false, true)) {
+        // a plan tread is working now
+        return;
+      }
+      try {
+        Thread planThread = new Thread(() -> {
+          int retry = 0;
+          boolean isHaveTask = false;
```

Comment:
Is `planSuccess` better here as a parameter? 
Or we should just check the size of `tasks` instead of create a new parameter.

Reply:
I removed the flag `isHaveTask`.